### ISSUE TITLE
Fail to do EVA repair on any building, check sunlight level

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#509]
**Describe the bug**
- Excessive messages regarding Repair bot not being able to go inside an ERV Base
 
**Expected behavior**
- It's not expected to send a Repair Bot to ERV base 1 to repair the *meteorite impact damage*.
- ERV Base 1 as an inhabitable building does not need the [INSIDE] type of repair.

**Screenshots**
![image](https://user-images.githubusercontent.com/1168584/142966659-cd6e9ab8-29f5-4e3e-bed0-ebc4550c7799.png)
ERV Base 1 sustains both damages that require both the [INSIDE] and [EVA] repair
![image](https://user-images.githubusercontent.com/1168584/142966966-42c1ff9e-85b1-439f-8813-b3f99bd6867c.png)

**Version **
  - build 6470

**Additional context**
```
00-Adir-10:304.365 (Warning) RepairInsideMalfunction [x29] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
00-Adir-10:351.197 (Warning) RepairInsideMalfunction [x70] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
00-Adir-10:397.013 (Warning) RepairInsideMalfunction [x32] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
00-Adir-10:444.028 (Warning) RepairInsideMalfunction [x54] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
00-Adir-10:489.014 (Warning) RepairInsideMalfunction [x75] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
00-Adir-10:543.965 (Warning) RepairInsideMalfunction [x65] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
00-Adir-10:597.394 (Warning) RepairInsideMalfunction [x66] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
00-Adir-10:643.072 (Warning) RepairInsideMalfunction [x222] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
00-Adir-10:697.839 (Warning) RepairInsideMalfunction [x150] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
00-Adir-10:744.993 (Warning) RepairInsideMalfunction [x40] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
00-Adir-10:791.547 (Warning) RepairInsideMalfunction [x46] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
00-Adir-10:837.778 (Warning) RepairInsideMalfunction [x45] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
00-Adir-10:884.563 (Warning) RepairInsideMalfunction [x29] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
00-Adir-10:925.396 (Severe) Settlement [x13] : [Hawthorne] Water Ration Level: 8016.0
00-Adir-10:929.133 (Warning) RepairInsideMalfunction [x44] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
00-Adir-10:985.700 (Warning) RepairInsideMalfunction [x34] : [Hawthorne] RepairBot 001 - Can not walk inside ERV Base 1
```

**Comment by mokun** [comment]
@bevans2000 ,

In the long run, it's good to allow a person to enter an inhabitable building and not just repairing the outside hull of a building.

What do you think ?

**Comment by bevans2000** [comment]
Yes I know of this problem. I have an idea how to resolve it.

**Comment by mokun** [comment]
> In the long run, it's good to allow a person to enter an inhabitable building and not just repairing the outside hull of a building.

The settlers are still not able to do the [EVA] type of repair on the building ERV-B 1 which is inhabitable and is barred from entry.

What should we do ?

![image](https://user-images.githubusercontent.com/1168584/144329744-4b6c67d0-f053-4dba-ba14-f6bdb030cc27.png)

```
00-Bora-07:549.329 (Info) Malfunction [x72] : [System] Repairer Teresa Green leaving the scene.
00-Bora-07:588.109 (Info) Farming : [New Eden] Inflatable Greenhouse 1 - Running out of soybean tissue-culture.
00-Bora-07:588.109 (Info) Crop : [New Eden] Inflatable Greenhouse 1 -  No soybean tissue-culture left. Restocking.
00-Bora-07:588.109 (Info) Farming : [New Eden - Inflatable Greenhouse 1] Jade Dube - Planted a new crop of soybean.
00-Bora-07:662.854 (Info) PerformLaboratoryExperiment : [New Eden] Alexia Ouellet - Just spent 18811.8 millisols in performing collaborative lab experiments on Medicine #3.
00-Bora-07:785.492 (Info) Malfunction [x204] : [System] Repairer Teresa Green leaving the scene.
00-Bora-08:016.579 (Info) Malfunction [x78] : [System] Repairer Jade Dube leaving the scene.
00-Bora-08:248.796 (Info) Malfunction [x225] : [System] Repairer Teresa Green leaving the scene.
00-Bora-08:479.560 (Info) Malfunction [x123] : [System] Repairer Teresa Green leaving the scene.
00-Bora-08:710.717 (Info) Malfunction [x142] : [System] Repairer Teresa Green leaving the scene.
00-Bora-08:944.918 (Info) Malfunction [x198] : [System] Repairer Alexia Ouellet leaving the scene.
00-Bora-09:215.016 (Info) Malfunction [x139] : [System] Repairer Rayan Brown leaving the scene.
00-Bora-09:266.022 (Info) ToggleFuelPowerSource [x72] : [New Eden] Alexia Ouellet - Turning on Fuel Power Source in Lander Hab 1.
00-Bora-09:542.141 (Info) Crop : [New Eden - Inflatable Greenhouse 1] GardenBot 001 - Harvested a total of 71.21 kg potato.
00-Bora-09:542.141 (Info) Crop : [New Eden - Inflatable Greenhouse 1] GardenBot 001 - Closed out the final harvest of potato.
00-Bora-09:542.141 (Info) ToggleFuelPowerSource [x55] : [New Eden] Alexia Ouellet - Turning off Fuel Power Source in Lander Hab 1.
00-Bora-09:542.141 (Info) Malfunction [x66] : [System] Repairer Teresa Green leaving the scene.
00-Bora-09:788.200 (Info) Malfunction [x74] : [System] Repairer Alexia Ouellet leaving the scene.
00-Bora-09:788.477 (Warning) EnterAirlock : [5.05 S 4.21 E] Rayan Brown - Could not depressurize EVA Airlock 1.
00-Bora-10:023.209 (Info) Malfunction [x219] : [System] Repairer Teresa Green leaving the scene.
00-Bora-10:264.678 (Info) Malfunction [x73] : [System] Repairer Rayan Brown leaving the scene.
00-Bora-10:344.475 (Info) Farming : [New Eden] Inflatable Greenhouse 1 - Running out of peanut tissue-culture.
00-Bora-10:344.475 (Info) Crop : [New Eden] Inflatable Greenhouse 1 -  No peanut tissue-culture left. Restocking.
00-Bora-10:344.475 (Info) Farming : [New Eden - Inflatable Greenhouse 1] Alexia Ouellet - Planted a new crop of peanut.
00-Bora-10:502.732 (Info) Malfunction [x109] : [System] Repairer Teresa Green leaving the scene.
00-Bora-10:735.134 (Info) Malfunction [x129] : [System] Repairer Teresa Green leaving the scene.
00-Bora-10:817.769 (Warning) EnterAirlock [x89] : [5.05 S 4.21 E] Rayan Brown - Could not depressurize EVA Airlock 1.
00-Bora-10:974.757 (Info) Malfunction [x96] : [System] Repairer Teresa Green leaving the scene.
00-Bora-10:981.470 (Config) Simulation : Autosaving the simulation as 2021-12-01T15.16.54_sol38_r6528.sim.
00-Bora-10:981.470 (Config) Simulation : heapSize: 384.00 MB    heapMaxSize: 1.50 GB    heapFreeSize: 245.30 MB
00-Bora-10:981.470 (Config) Simulation :            File size : 1.25 MB
00-Bora-10:981.470 (Config) Simulation : Done saving. The simulation resumes.
00-Bora-11:121.202 (Warning) EnterAirlock : [5.05 S 4.21 E] Alexia Ouellet - Could not depressurize EVA Airlock 1.
00-Bora-11:189.372 (Info) Malfunction [x318] : [System] Repairer Alexia Ouellet leaving the scene.
00-Bora-11:425.189 (Info) Malfunction [x95] : [System] Repairer Rayan Brown leaving the scene.
00-Bora-11:651.962 (Info) Crop : [New Eden - Inflatable Greenhouse 1] GardenBot 001 - Harvested a total of 62.95 kg carrot.
00-Bora-11:651.962 (Info) Crop : [New Eden - Inflatable Greenhouse 1] GardenBot 001 - Closed out the final harvest of carrot.
00-Bora-11:656.022 (Info) Malfunction [x72] : [System] Repairer Teresa Green leaving the scene.
00-Bora-11:896.383 (Info) Malfunction [x146] : [System] Repairer Rayan Brown leaving the scene.
00-Bora-12:127.955 (Info) Malfunction [x122] : [System] Repairer Alexia Ouellet leaving the scene.
00-Bora-12:256.175 (Info) Farming : [New Eden] Inflatable Greenhouse 1 - Running out of peanut tissue-culture.
00-Bora-12:256.175 (Info) Crop : [New Eden] Inflatable Greenhouse 1 -  No peanut tissue-culture left. Restocking.
00-Bora-12:256.175 (Info) Farming : [New Eden - Inflatable Greenhouse 1] Alexia Ouellet - Planted a new crop of peanut.
00-Bora-12:361.464 (Info) Malfunction [x222] : [System] Repairer Teresa Green leaving the scene.
00-Bora-12:594.628 (Info) Malfunction [x81] : [System] Repairer Teresa Green leaving the scene.
```

**Comment by mokun** [comment]
Something is stopping Teresa and others from exiting the airlock to walk over there.

![image](https://user-images.githubusercontent.com/1168584/144330611-7a928465-b2da-4679-8a94-86f06d1a1245.png)

**Comment by mokun** [comment]
I was wrong. They did exit the airlock without any issues at all.

But after they got out, they "instantly" walked back to the airlock. 

This is due to the fact that ERV building is inhabitable and there is no code to tell a person to approach it and stay right outside and begin the [EVA] type of repair on its outer frame.


![image](https://user-images.githubusercontent.com/1168584/144331211-68e5e246-de0d-44f3-a31a-164fc967cd77.png)

**Comment by bevans2000** [comment]
I'm looking at Walking as part of #470 so I can have a look. However I've found that recently a lot kf people fail to use the Airlock because they are not fit enough. It seems we have changed the fitness check?

**Comment by mokun** [comment]
> recently a lot kf people fail to use the Airlock because they are not fit enough. It seems we have changed the fitness check?

hmm I don't recall changing anything regarding fitness. 

But I did streamline the reservation system. 

I'm looking at the task class `RepairEVAMalfunction`.

Just added a logger statement to see if calling the method `RepairHelper.hasRepairParts(containerUnit, malfunction)` returns false. 

The culprit is perhaps the fact that the repair parts are not available.
 
```java
	/**
	 * Perform the repair malfunction phase of the task.
	 *
	 * @param time the time to perform this phase (in millisols)
	 * @return the time remaining after performing this phase (in millisols)
	 */
	private double repairMalfunctionPhase(double time) {
		boolean endTask = isDone();

		// Check for radiation exposure during the EVA operation.
		endTask |= (isRadiationDetected(time)
					|| malfunction.isWorkDone(MalfunctionRepairWork.EVA));
		endTask |= malfunction.isWorkDone(MalfunctionRepairWork.EVA);
		endTask |= (shouldEndEVAOperation() || addTimeOnSite(time));
		endTask |= (!person.isFit());
		if (endTask) {
			// Return all the time
	        if (worker.isOutside()) {
	        	setPhase(WALK_BACK_INSIDE);
	        }
	        else {
	    		endTask();
	        }
    		return time;
        }

		double workTime = 0;
		if (person != null) {
			workTime = time;
		} else {
			// A robot moves slower than a person and incurs penalty on workTime
			workTime = time / 2;
		}

		// Determine effective work time based on "Mechanic" skill.
		int mechanicSkill = worker.getSkillManager().getEffectiveSkillLevel(SkillType.MECHANICS);

		if (mechanicSkill == 0)
			workTime /= 2;
		if (mechanicSkill > 1)
			workTime += workTime * (.2D * mechanicSkill);

		if (RepairHelper.hasRepairParts(containerUnit, malfunction)) {
			RepairHelper.claimRepairParts(containerUnit, malfunction);
		}

		else {
			logger.info(worker, "Repair parts not available for "
					+ malfunction + " from " + containerUnit.getName() + ".");

        	if (worker.isOutside()) {
        		setPhase(WALK_BACK_INSIDE);
        	}
        	else {
        		endTask();
        	}
            return time;
		}

		// Add EVA work to malfunction.
		double workTimeLeft = 0D;
		if (!malfunction.isWorkDone(MalfunctionRepairWork.EVA)) {
			workTimeLeft = malfunction.addWorkTime(MalfunctionRepairWork.EVA, workTime, worker.getName());
		}

		// Add experience points
		addExperience(time);

		// Check if an accident happens during repair.
		checkForAccident(time);

		// Check if there are no more malfunctions.
		if (malfunction.isWorkDone(MalfunctionRepairWork.EVA)) {
			logger.info(worker, "Wrapped up the EVA of " + malfunction.getName()
					+ " in "+ entity + " ("
					+ Math.round(malfunction.getCompletedWorkTime(MalfunctionRepairWork.EVA)*10.0)/10.0 + " millisols spent).");
            if (worker.isOutside()) {
            	setPhase(WALK_BACK_INSIDE);
            }
            else {
        		endTask();
            }
		}

		return workTimeLeft;
	}
```

**Comment by mokun** [comment]
@bevans2000 

We should probably at least offer a quick fix on this before releasing v3.4.0. 

What are your thoughts on how to address it ?

**Comment by mokun** [comment]
> Also it appears we haven't written any codes telling a person how to get near a building that has no Life Support inside and simply stay right outside and begin the [EVA] type of repair on its outer frame.

Could this be the issue ? 

They got "turned away" immediately after they arrived at an inhabitable building ?

**Comment by mokun** [comment]
In RepairEVAMalfunction, we have the following: 

```java
public class RepairEVAMalfunction extends EVAOperation implements Repair, Serializable {

	/** default serial id. */
	private static final long serialVersionUID = 1L;

	/** default logger. */
	private static final SimLogger logger = SimLogger.getLogger(RepairEVAMalfunction.class.getName());

	/** Task name */
	private static final String NAME = Msg.getString("Task.description.repairEVAMalfunction"); //$NON-NLS-1$

	/** Task phases. */
	private static final TaskPhase REPAIRING = new TaskPhase(Msg.getString("Task.phase.repairing")); //$NON-NLS-1$

	// Data members
	/** The malfunctionable entity being repaired. */
	private Malfunctionable entity;

	/** The malfunction to be repaired. */
	private Malfunction malfunction;

	/** The container unit the person started the mission in. */
	private Unit containerUnit;

	public RepairEVAMalfunction(Person person) {
		super(NAME, person, true, 25, SkillType.MECHANICS);

....

		// Start if found
		if (entity != null) {
			setDescription(Msg.getString("Task.description.repairEVAMalfunction.detail", malfunction.getName(),
					entity.getNickName())); // $NON-NLS-1$

			// Determine location for repairing malfunction.
			setOutsideLocation((LocalBoundedObject) entity);

			// Can fail to get a path and Task will be Done
			if (!isDone()) {
	            if (person.isInside()) {
	            	setPhase(WALK_TO_OUTSIDE_SITE);
	            }

				RepairHelper.startRepair(malfunction, person, MalfunctionRepairWork.EVA, entity);

				// Initialize phase
				addPhase(REPAIRING);
			}
		}
		else {
        	if (person.isOutside())
        		setPhase(WALK_BACK_INSIDE);
        	else
        		endTask();
		}
	}

```

The question is, what happen when the person arrives, will it stay right outside the building and perform the repair ? 

In EVAOperation, it shows: 

```java
	/**
	 * Perform the walk to outside site phase.
	 *
	 * @param time the time to perform the phase.
	 * @return remaining time after performing the phase.
	 */
	private double walkToOutsideSitePhase(double time) {
	    // If not at field work site location, create walk outside subtask.
        if (person.isInside()) {// || !closeToLocation) {
        	// A person is walking toward an airlock or inside an airlock
            Walk walkingTask = Walk.createWalkingTask(person, outsideSitePos, 0, null);
            if (walkingTask != null) {
                addSubTask(walkingTask);
            }
            else {
				logger.severe(person, "Cannot walk to outside site.");
                endTask();
            }
        }
        else {
        	// In case of DigLocalRegolith, set to task phase COLLECT_REGOLITH
            setPhase(getOutsideSitePhase());
        }

        return time;
    }
```

**Comment by bevans2000** [comment]
The method ```setOutsideLocation``` determines the position outside near the broken building. 
What exactly are the symptoms? If the Building is inhabitable then only EVA repair is needed. A Robot doesn't look for EVA repairs so the problem should be fixed .

**Comment by mokun** [comment]
> What exactly are the symptoms? 

A bunch of people going out but they are NOT able to add any repair time. So after days, it's still `0/128 millisols`.

**Comment by bevans2000** [comment]
When I have seen this it is because there is an Airlock issue. I can give it a try tonight on my branch.

**Comment by mokun** [comment]
Here I have an incident in that the Meteorite damage is on Lander Hab 3.

![image](https://user-images.githubusercontent.com/1168584/146467550-9a7e4a2f-674f-4768-9fa3-472d5f4d218b.png)

As you can see, the [INSIDE] repair was done (159/159), but the [EVA] repair could not even be initiated (0/126).

I wonder why.

**Comment by mokun** [comment]
@bevans2000 

Let's find out why no one can do [EVA]-type of repair on a Lander Hab.  

There's no problem with [INSIDE]-type of repair.

**Comment by bevans2000** [comment]
Yes this is odd. Been super busy with with but might have some time over the weekend to look into this.

**Comment by bevans2000** [comment]
This relates to an issue I previously raised in 4c2b5e8e9817b9c2f10d3cfbb8989be6defdc263. ```EVAOperation``` can not complete because the ```isGettingDark``` method is returning true every time.

**Comment by mokun** [comment]
In this [commit](https://github.com/mars-sim/mars-sim/commit/4c2b5e8e9817b9c2f10d3cfbb8989be6defdc263), I removed getTrend() and revise isGettingDark() method. 

Guess I didn't do it right. But before revising it, the EVA repair has NOT been working.

So, let's fix both this issue and whatever stopping the EVA repair.

**Comment by bevans2000** [comment]
All I can say is last night when I debugged it all the cases stopped the EVA immediately because it was getting dark. People moved to the outside site and returned straight away.
A while ago there was the airlock problem which may have been a problem.

**Comment by bevans2000** [comment]
This may be related to the "tooDark EVA" problem c1fcf6339994929a34a200e0c6ee439454a44acd

**Comment by mokun** [comment]
I just found out this has nothing to do with that tooDark EVA issue.

See below :
![image](https://user-images.githubusercontent.com/1168584/173444423-1d94e4eb-350e-4c50-80d2-5bf86fb834ce.png)

Note that the [Inside] Repair has been done in Lounge 1 in S Point but the [EVA] repair never started. In total, it's about 51% done. 

![image](https://user-images.githubusercontent.com/1168584/173444973-804adf2b-e9a9-417f-bef0-e7aa3d24c28c.png)

**Comment by mokun** [comment]
After a while, most settlements will have suffered from Meteorite Damages

![image](https://user-images.githubusercontent.com/1168584/173446985-f668a6b3-1709-4d59-acd3-7d3c8997c860.png)

It seems that the repair parts are not available. In particular, most settlements don't have plastic sheets or steel sheet available.

If that's the case, then the base value point for these two parts need to bump up higher to prompt for actions.

But

**Comment by mokun** [comment]
In another scenario, Lena went out to do the EVA repair.

![image](https://user-images.githubusercontent.com/1168584/173489562-9618d2ff-b0ab-4738-aba4-555fe9797b31.png)

This time, there is no needed parts for repair. Since a hole or a puncture will result is on the outer wall, should it have at least required one part for the repair ?

Also, the EVA repair stays at 12/129 millisols for a long time, while the percentage is going up from 60%+ to 95%. Why does it stop at 12/129 ?

**Comment by mokun** [comment]
@bevans2000 

I tried manually start other malfunctions. It looks at the black tooltip is unable to update itself when more millisols of work are done.

**Comment by mokun** [comment]
In another Meteorite Damage repair, the building panel of the Astronomy Observatory 1 building shows only [EVA] type of repair and never shows the [Inside] type of repair.

![image](https://user-images.githubusercontent.com/1168584/173492151-d593f9e2-3da1-484a-8066-8a01528fc241.png)

Again the black tooltip doesn't get the millisols updated.

**Comment by bevans2000** [comment]
That is how the repair logic for parts has always worked. Are you saying the tooltip is not updating?
There are too many comments in different problems on this issue. We should close this and create more specific ones.

**Comment by bevans2000** [comment]
Essentially the UI has a lot of caching and cannot be trusted sometimes if the rendered data looks wrong. The Console commands will always give the true state of things because there is no caching nor presentation logic.

**Comment by mokun** [comment]
> Essentially the UI has a lot of caching and cannot be trusted sometimes if the rendered data looks wrong.

Let's take a look at how a meteorite damage repair proceeds. Or you can manually generate one.

Then you can use console to examine whether the EVA repair is properly done and why not.

**Comment by mokun** [comment]
One problem we have is that from time to time they requires certain needed parts for an EVA repair. 

However, these parts are NOT readily available in that settlement. That's why the repair is halted.

So I recalled in some commit I specifically removed those rarely made parts from malfunctions.xml.

But in the long term, it's really about fixing the supply chain.

In our case, how a repair would trigger something to make the VP of a needed part go sky-rocket so that the manufacturing process will pick up the urgent need of producing that part.

In conclusion, there are probably multiple things to fix this **supply chain** issue, much like the so called *chip shortage* we have in the world right now as we speak.

**Comment by bevans2000** [comment]
It also relates to #506 & #601 since to corrrecty do these changes it needs to be based on demand which relates to Value Point.

**Comment by mokun** [comment]
It's still happening as of now. I just witnessed another one. 

It's a 4-person SpaceX settlement. 

Meteorite Impact damage was on **Atmospheric Processor 1**. 

It shows `None` on needed repair parts.

The indicator bar stops at 67% forever.

3 out of 4 persons have started the Repair EVA task. They went outside through the airlock but never stop next to the **Atmospheric Processor 1**

The trouble is as an inhabitable building it's NOT supposed to have `Inside repair time`. It even said 0 millisols.  

Q: what does it mean by `Inside repair time: 0 millisols` ?

Can you change it to `0 / 50 millisols`, so as to display both the *done* and *required* millisols ? 

![image](https://user-images.githubusercontent.com/1168584/182275029-40a81889-426d-49cd-9ec9-c72fd62327e9.png)

**Comment by mokun** [comment]
Interestingly, in Settlement Window, Maintenance tab, things are better.

![image](https://user-images.githubusercontent.com/1168584/182275921-748ef513-9300-4d99-ba8c-0e3d328d1169.png)

It shows clearly `EVA Repair Time: 49/73 millisols`. There's no `Inside Repair Time` listed. 

So, I'll replicate the change that have been made on `TabPanelMaintenance` and do it on `BuildingPanelMaintenance`.

**Comment by bevans2000** [comment]
You should create a common tooltip class to render Repairs in both cases.

**Comment by mokun** [comment]
> You should create a common tooltip class to render Repairs in both cases.

Yea the progress bar, the panels and the text around it are almost the same in both cases, except for a few things. In that previous commit, I've renamed the elements in both cases to use the same verbiage so that I can keep track of things better. 

But in terms of the Settlement Window, I have wanted to first create a newtab called `TabPanelMalfunction` by splitting the codes from `TabPanelMaintenance` 

It's because in Building Window, there's already two separate tabs, namely, `BuildingPanelMalfunctionable` and `BuildingPanelMaintenance`.

**Comment by bevans2000** [comment]
Something has gone wrong with the Airlock logic again. I've got an airlock with 1 person coming in and 2 going out. and it's stuck in a loop.
![Screenshot 2022-08-05 225532](https://user-images.githubusercontent.com/20720166/183216415-d7ffbcfe-df27-4ee4-bc21-acf25a7ff95a.png)

**Comment by mokun** [comment]
So, in your case, Scott has no suit while ingressing. I wonder at what task phase he's stuck at ? Has he just doffed off his suit ?

**Comment by mokun** [comment]
> 2 going out.

I've found something and fixing it.

I'll be careful with the use of inAirlock() and exitAirlock() when a person is in a limbo state of transition from inside an airlock to outside.

**Comment by mokun** [comment]
If a person is in the process of doffing off his EVA suit, we  must make sure the airlock will not change its state of being pressurized. 

Or else, right that after the person has taken off the suit, the person cannot go out.

**Comment by bevans2000** [comment]
Sorry I didn't capture what they were doing. I will do it if it reoccurs.

**Comment by mokun** [comment]
isEnoughSunlightForEVA() in EVAOperation is good by itself.

However, finding out if the sky is getting darker or the sun is setting is different.

Having the amount of sunlight at least `minEVASunlight` still won't tell us if it's sunrise or sunset.

Now I've figured out a way to gauge if it's sunrise or sunset.

Fixing the issue with isGettingDark() once and for all right now.

**Comment by bevans2000** [comment]
Ok. Keep it all inside isEnoughtSunlightFieEVA because that method hives a single way to judge the light and is called from numerous places. We need it to be consistent in the approach

**Comment by mokun** [comment]
It all looks good now. Closing it.
